### PR TITLE
Cancel pending merges when a key is deleted

### PR DESCRIPTION
**Pull request by sosek108**
### Details

`Onyx.merge` could resurrect a key that a concurrent deletion had removed, bringing back its **pre-deletion contents**.

`Onyx.merge` applies its delta to a value captured by an async `OnyxUtils.get(key)`. `Onyx.set` and `Onyx.multiSet` cancel merges queued before them so the parked callback aborts. `setCollection`, `mergeCollection` and `partialSetCollection` never did, so an in-flight merge wrote its stale snapshot back with the delta on top. After `cache.drop()` the `hasCacheForKey` re-read from #817 cannot help, because the key is no longer cached.

Two fixes:

1. The three collection writers cancel queued merges at their synchronous entry, before any `await`, so they discard only merges issued before the deleting call. Cancelling in the `cache.drop()` loop instead would also kill merges issued after the removal, which must win.
2. The merge abort guard compares queue identity instead of nullishness, so a merge whose queue was cancelled and then repopulated by a later merge no longer applies that merge's deltas to its own stale snapshot.

### Related Issues
https://github.com/Expensify/App/issues/98073

### Linked E/App PR

https://github.com/Expensify/App/pull/101171

### Automated Tests

Five tests in `tests/unit/onyxTest.ts`, added to the existing `merge > concurrency with Onyx.update` block. Each one parks `OnyxUtils.get` with `createDeferredTask` so the deletion lands first, and checks both the cache and storage.

| Test | Deletion path |
|---|---|
| `should keep the key deleted when a pending merge read resolves after Onyx.update deleted it` | `partialSetCollection` |
| `should keep the key deleted when a pending merge read resolves after mergeCollection deleted it` | `mergeCollectionWithPatches` |
| `should keep the key deleted when a pending merge read resolves after setCollection dropped it` | `setCollectionWithRetry` |
| `should not resurrect pre-deletion data when a merge is queued after the key was removed` | abort guard identity |
| `should keep the key deleted when a single-key Onyx.update removal lands during a pending merge` | `set` |

### Manual Tests

The race is not manually reproducible without the instrumentation, so there are no direct reproduction steps to verify if the fix is working - we rely on unit testing. I propose [these steps](https://github.com/Expensify/App/issues/97400) as regression verification:

1. Go offline.
2. Open a report with a split expense.
3. Delete the split expense.
4. Verify it does not reappear.
5. Go back online and verify it stays deleted after the queued updates flush.

### Author Checklist

- [ ] I linked the correct issue in the `### Related Issues` section above
- [ ] I linked the corresponding Expensify/App PR in the `### Linked E/App PR` section above, and verified this change against it (E/App CI passed and manual testing completed)
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

### CODE REVIEW

**Comment by fabioh8010** [comment]
@codex review

**Comment by sosek108** [comment]
@codex review

**Comment by sosek108** [comment]
@MelvinBot review

File: lib/OnyxUtils.ts
Comment (grgia) [inline lib/OnyxUtils.ts:1573]: NAB. should `multiSetWithRetry` also call `cancelPendingMergesForKey` for `keysToRemove`? 

Data removed with Onyx.multiSet could come back if a merge on the same key was still in progress.

**Review by grgia — COMMENTED** [review]
NAB

**Comment by mountiny** [comment]
@mkhutornyi can you please test the chnages in App? thanks

File: lib/OnyxUtils.ts
Comment (sosek108) [reply lib/OnyxUtils.ts:1573]: Good catch, it was. Fixed in 68a9b7e1.

**Review by grgia — DISMISSED** [review]

**Comment by mountiny** [comment]
@mkhutornyi how is it looking?

**Comment by mkhutornyi** [comment]
Almost done

File: tests/unit/onyxTest.ts [hunk 1]
```diff
@@ -2723,6 +2723,363 @@ describe('Onyx', () => {
...
+
+                deferredGet.resolve();
+                await mergePromise;
+                await waitForPromisesToResolve();
+
+                expect(cache.get(member1)).toEqual({itemA: {id: 'a', childID: '1'}});
+                expect(await StorageMock.getItem(member1)).toEqual({itemA: {id: 'a', childID: '1'}});
+            });
+
+            it('should not cancel a merge that was queued after Onyx.clear() was called', async () => {
```
Comment (mkhutornyi) [inline tests/unit/onyxTest.ts:3013]: ### `clear()` still cancels merges queued after it when a merge for that key is already in flight

`cancelPendingMerges` guards on array reference identity, but [`merge()`](https://github.com/Expensify/react-native-onyx/blob/68a9b7e18e8ad9e3039e86910f458704887dd05d/lib/Onyx.ts#L252-L254) appends to the existing array instead of creating a new one when a merge for that key is already pending:

```js
if (mergeQueue[key]) {
    mergeQueue[key].push(changes);   // same array -> identity still matches
    return mergeQueuePromise[key];
}
```

So the guard in [`cancelPendingMerges`](https://github.com/Expensify/react-native-onyx/blob/68a9b7e18e8ad9e3039e86910f458704887dd05d/lib/OnyxUtils.ts#L937-L943) still matches and the post-`clear()` delta is dropped, while its promise resolves successfully and nothing is logged.

This is the new test case with the second merge moved *before* the first one drains:

1. Park `member1`'s `OnyxUtils.get` and `OnyxUtils.getAllKeys` (same mocks as that test).
2. `const firstMergePromise = Onyx.merge(member1, {itemA: {id: 'a'}});` — creates array `qc1`.
3. `const clearPromise = Onyx.clear();` then `await waitForPromisesToResolve()` — captures `[[member1, qc1]]` and parks on `getAllKeys()`.
4. `const secondMergePromise = Onyx.merge(member1, {itemB: {id: 'b'}});` — **without** resolving `firstGet` first, so `{itemB}` is pushed onto `qc1`.
5. `deferredGetAllKeys.resolve(); await clearPromise;` — `mergeQueue[member1] === qc1`, so the whole array is cancelled.
6. `firstGet.resolve(); await Promise.all([firstMergePromise, secondMergePromise]);`

**Expected:** `cache.get(member1)` → `{itemB: {id: 'b'}}`
**Actual:** `cache.get(member1)` → `undefined`, storage `null`. Both merges dropped, both promises resolved.

The existing test passes only because it does `firstGet.resolve(); await firstMergePromise;` before issuing the second merge, which drains the queue and hands the second merge a fresh array.

----

Btw I found no user visible bug for this.

**Comment by mkhutornyi** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified that the composer does not automatically focus or open the keyboard on mobile unless explicitly intended. This includes checking that returning the app from the background does not unexpectedly open the keyboard.
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: HybridApp
    - [x] Android: mWeb Chrome
    - [x] iOS: HybridApp
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: HybridApp</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: HybridApp</summary>


https://github.com/user-attachments/assets/960c6f54-2e3d-41a6-80d1-9887ca7e7fe4



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/user-attachments/assets/ba25074c-0b18-4f14-bcce-420efc9e0dee


</details>

**Review by mkhutornyi — APPROVED** [review]

File: tests/unit/onyxTest.ts (hunk 1 above)
Comment (sosek108) [reply tests/unit/onyxTest.ts:3013]: Added new test case and fixed this in https://github.com/Expensify/react-native-onyx/pull/846/changes/0eed0fd1fa995a4326066140d34b596823906fff 

Fix:
- lib/OnyxUtils.ts: getPendingMergeEntries now also records each queue's length when clear() captures it. In cancelPendingMerges:
  - If the queue didn't grow, it's cancelled as before.
  - If it grew, only the pre-clear() changes are removed, and the queue is flagged in a WeakSet (hasStaleMergeRead).
- lib/Onyx.ts: in the merge callback, a flagged queue uses undefined as its base instead of the stale valueFromGet. The cache still takes priority, as before.

File: lib/Onyx.ts [hunk 2]
```diff
@@ -253,18 +253,20 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
...
+            if (mergeQueue[key] !== queuedChanges) {
                 return Promise.resolve();
             }
 
             // Other writers (notably Onyx.update's mergeCollection path, which doesn't participate in mergeQueue)
             // can land between get() resolving and this callback running. Applying the delta on top of the value
             // captured back then and broadcasting it would overwrite those writes wholesale, so re-read the cache.
-            const existingValue = cache.hasCacheForKey(key) ? (cache.get(key) as OnyxInput<TKey> | undefined) : valueFromGet;
+            const baseValue = OnyxUtils.hasStaleMergeRead(queuedChanges) ? undefined : valueFromGet;
+            const existingValue = cache.hasCacheForKey(key) ? (cache.get(key) as OnyxInput<TKey> | undefined) : baseValue;
```
Comment (grgia) [inline lib/Onyx.ts]: NAB. the real `OnyxUtils.get` calls `cache.set(key, val)` with the pre-clear value when it resolves, so `hasCacheForKey` beats `hasStaleMergeRead` here. the test only passes because `get` is mocked.

**Review by grgia — APPROVED** [review]
Not sure we need to block on this @sosek108 

the edge case would be after signing out, data from the old session can come back for a key that was being merged while the clear ran.

**Comment by grgia** [comment]
Please address the above comment and let me know if we should still merge
